File: Graphs/P018.py
def brute(startWord,targetWord,wordList):
    wordSet=set(wordList)
    minStep=[float('inf')]
    def diff_by_oneword(w1,w2):
        diff=0
        for i in range(len(w1)):
            if w1[i]!=w2[i]:
                diff+=1
            if diff>1:
                return False
        return diff==1
    def dfs(word,step,visited):
        if word==targetWord:
            minStep[0]=min(minStep[0],step)
            return
        for next in wordList:
            if diff_by_oneword(next,word) and next not in visited:
                visited.add(next)
                dfs(next,step+1,visited)
                visited.remove(next)
    dfs(startWord,1,set([startWord]))
    return minStep[0] if minStep[0]!=float('inf') else 0

from collections import deque
def optimal(startword,targetword,wordList):
    wordSet=set(wordList)
    q=deque([(startword,1)])
    while q:
        word,step=q.popleft()
        if word==targetword:
            return step
        for i in range(len(word)):
            for ch in 'abcdefghijklmnopqrstuvwxyz':
                newWord=word[:i]+ch+word[i+1:]
                if newWord in wordSet:
                    q.append((newWord,step+1))
                    wordSet.remove(newWord)
    return 0

startWord = "hit"
targetWord = "cog"
wordList = ["hot", "dot", "dog", "lot", "log", "cog"]
print("Brute", brute(startWord, targetWord, wordList))
print("Optimal :", optimal(startWord, targetWord, wordList))

File: Striver_C++_Problems/Graphs/P018.cpp
// P018.cpp - Word Ladder 1 (shortest transformation length)

#include <bits/stdc++.h>
using namespace std;

bool diff_by_one(const string& a, const string& b) {
    int diff = 0;
    for (int i = 0; i < (int)a.size(); i++) {
        if (a[i] != b[i]) {
            diff++;
            if (diff > 1) return false;
        }
    }
    return diff == 1;
}

int brute(const string& startWord, const string& targetWord,
          const vector<string>& wordList) {
    unordered_set<string> wordSet(wordList.begin(), wordList.end());
    int INF = INT_MAX;
    int minStep = INF;

    function<void(string,int,unordered_set<string>&)> dfs =
        [&](string word, int step, unordered_set<string>& visited) {
            if (word == targetWord) {
                minStep = min(minStep, step);
                return;
            }
            for (auto &next : wordList) {
                if (!visited.count(next) && diff_by_one(word, next)) {
                    visited.insert(next);
                    dfs(next, step + 1, visited);
                    visited.erase(next);
                }
            }
        };

    unordered_set<string> visited;
    visited.insert(startWord);
    dfs(startWord, 1, visited);
    return (minStep == INF ? 0 : minStep);
}

// Optimal: BFS
int optimal(const string& startWord, const string& targetWord,
            const vector<string>& wordList) {
    unordered_set<string> wordSet(wordList.begin(), wordList.end());
    if (!wordSet.count(targetWord)) return 0;
    queue<pair<string,int>> q;
    q.push({startWord, 1});

    while (!q.empty()) {
        auto [word, step] = q.front(); q.pop();
        if (word == targetWord) return step;
        for (int i = 0; i < (int)word.size(); i++) {
            string temp = word;
            for (char ch = 'a'; ch <= 'z'; ch++) {
                temp[i] = ch;
                if (wordSet.count(temp)) {
                    q.push({temp, step + 1});
                    wordSet.erase(temp);
                }
            }
        }
    }
    return 0;
}

int main() {
    string startWord = "hit";
    string targetWord = "cog";
    vector<string> wordList = {"hot","dot","dog","lot","log","cog"};

    cout << "Brute "   << brute(startWord, targetWord, wordList) << "\n";
    cout << "Optimal " << optimal(startWord, targetWord, wordList) << "\n";
    return 0;
}
